vaadin-upload-flow-parent/vaadin-upload-flow/src/main/java/com/vaadin/flow/component/upload/Upload.java

```diff
@@ -560,17 +561,48 @@ public void setI18n(UploadI18N i18n) {
 
         runBeforeClientResponse(ui -> {
             if (i18n == this.i18n) {
-                JsonObject i18nObject = (JsonObject) JsonSerializer
+                JsonObject i18nJson = (JsonObject) JsonSerializer
                         .toJson(this.i18n);
-                for (String key : i18nObject.keys()) {
-                    ui.getPage().executeJavaScript(
-                            "$0.set('i18n." + key + "', $1)", getElement(),
-                            i18nObject.get(key));
-                }
+
+                // Remove null values so that we don't overwrite existing WC
+                // translations with empty ones
+                deeplyRemoveNullValuesFromJsonObject(i18nJson);
+
+                // Assign new I18N object to WC, by deeply merging the existing
+                // WC I18N, and the values from the new UploadI18N instance,
+                // into an empty object
+                getElement().executeJs(
+                        "const dropFiles = Object.assign({}, this.i18n.dropFiles, $0.dropFiles);"
+                                + "const addFiles = Object.assign({}, this.i18n.addFiles, $0.addFiles);"
+                                + "const error = Object.assign({}, this.i18n.error, $0.error);"
+                                + "const uploadingStatus = Object.assign({}, this.i18n.uploading.status, $0.uploading && $0.uploading.status);"
+                                + "const uploadingRemainingTime = Object.assign({}, this.i18n.uploading.remainingTime, $0.uploading && $0.uploading.remainingTime);"
+                                + "const uploadingError = Object.assign({}, this.i18n.uploading.error, $0.uploading && $0.uploading.error);"
+                                + "const uploading = {"
+                                + "  status: uploadingStatus,"
+                                + "  remainingTime: uploadingRemainingTime,"
+                                + "  error: uploadingError," + "};"
+                                + "const units = $0.units || this.i18n.units;"
+                                + "this.i18n = Object.assign({}, this.i18n, $0, {"
+                                + "  addFiles: addFiles,"
+                                + "  dropFiles: dropFiles,"
+                                + "  uploading: uploading," + "  units: units,"
+                                + "});",
+                        i18nJson);
```

Comment:
Couldn't identify an easier / safer way to do a deep merge. Considered:
- importing a merge function from a library like lodash - that might have too big of an impact for this change alone
- copying a merge function from Github / SO - might have bugs or not work as expected in this use case (e.g. merging `units` array). Also would be terrible to put such a complex function inline into the Java code

Still, I'm open for other ideas on improving this. I tried to make the tests rather extensive to make up for this.

Note that the `units` array is not merged. If the UploadI18N config specifies one, it will always replace the one from the web component, even if the number of elements don't match. However that is the same behaviour as we have today.